### PR TITLE
Feature: posting cash payments

### PR DESCRIPTION
**Pull request by jniles**
This PR implements posting on the `cash` and `cash_item` tables as MySQL procedures similar to those for `invoice`.

**Notable Changes and Additions**
1. I've split up functions and procedures into `functions.sql` and `procedures.sql`, respectively.  The `procedures.sql` was approaching 1000 lines last I counted so we may need additional subdivisions.
2. `GetExchangeRate(enterpriseId, currencyId, date)` is now a database function!
3. Both cautions (prepay) and cash payments are supported via the `is_caution` flag.

**How it Works**
The `cash` and `cash_item` tables record exactly what happened in history.  If the client paid 4,500 FC against invoice A and 1,200 FC against invoice B, the `cash` table will have the `currency_id` for FCs and 5,700 as the `amount`.  The `amount` in each cash_item will be 4,500 and 1,200.  If the client made a caution payment, there are no `cash_item` lines.

_Gain/Loss on rounding is only done at posting time_

When the `PostCash(cashUuid)` procedure is called the following steps take place:
1. Posting metadata is gathered (`PostingUtil()`).
2. The exchange rate is calculated (`GetExchangeRate()`).
3. The cashbox's account is debited the (exchanged) `cash` amount.
4. **Caution:**  The debtor is credited the (exchange) `cash` amount.
5. **Cash Payment:** For each cash payment:
   1. The debtor is credited the (exchanged) `cash_item` amount.
   2. The `cash_item.invoice_uuid` is put in `reference_uuid`.
6. **Cash Payment:** The total amount owed on the invoices the client is currently paying is calculated as `previousInvoiceBalances`.
   1. If the difference between the amount paid and `previousInvoiceBalances` is greater than the payment currency's `minMonentaryUnit`, do nothing.  The debtor has not paid the full amount.
   2. If the difference is smaller than the `minMonentaryUnit`, the debtor has paid the full amount and the enterprise either gains or loses some money.
      1. If the difference is positive, the debtor overpaid slightly.  The enterprise `gain_account_id` is credited and the debtor is debited the difference.  The final invoice paid is assigned to the `reference_uuid`.
      2. If the difference is negative, the debtor underpaid slightly.  The enterprise `loss_account_id` is debited and the debtor is credited the difference.  The final invoice paid is assigned to the `reference_uuid`.

I am sure there will be bugs in this implementation, but now testing can begin!

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

File: client/src/partials/cash/cash.html [hunk 1]
```diff
@@ -50,7 +50,7 @@
                 <bh-currency-select
                   currency-id="CashCtrl.payment.currency_id"
                   validation-trigger="CashVoucherForm.$submitted"
-                  on-change="CashCtrl.digestExchangeRate()"
+                  on-change="CashCtrl.digestExchangeRate(currency)"
                   disable-ids="CashCtrl.disabledCurrencyIds"
                   cashbox-id="CashCtrl.cashbox.id"
```
Comment (DedrickEnc) [inline client/src/partials/cash/cash.html:55]: Why a component responsible of changing currency must be some how directly linked to the cash box?

File: client/src/partials/cash/cash.html [hunk 2]
```diff
@@ -110,7 +110,7 @@
 
                       <!-- only show this segment if using a foreign currency -->
                       <tr>
-                        <td colspan="2" class="text-right"><u>{{ "TABLE.COLUMNS.EXCHANGE_RATE" | translate }}</u></td>
+                        <td colspan="2" class="text-right"><b>{{ "TABLE.COLUMNS.EXCHANGE_RATE" | translate }}</b></td>
```
Comment (DedrickEnc) [inline client/src/partials/cash/cash.html:113]: Don't we have a CSS class to bold text?

File: client/src/partials/cash/cash.js [hunk 3]
```diff
@@ -29,15 +29,14 @@ function CashController(Cash, Cashboxes, AppCache, Currencies, Exchange, Session
...
-  // if there is no URL id but one in localstorage, the state with the
+  // if there is no URL id but one in localstorage, update the state with the
   // localstorage params
   } else if (cashboxId && !$state.params.id) {
     $state.go('^.window', { id : cashboxId }, { location: 'replace'});
   }
 
   // bind methods
   vm.openInvoicesModal = openInvoicesModal;
-  vm.openTransferModal = openTransferModal;
```
Comment (DedrickEnc) [inline client/src/partials/cash/cash.js:40]: It seems like the transfer process is no longer a part of cash box module? why?

File: client/src/partials/cash/modals/invoices.modal.js [hunk 4]
```diff
@@ -27,26 +26,34 @@ function CashInvoiceModalController(Debtors, debtorId, invoiceIds, ModalInstance
...
   };
 
+  function selectionChangeCallback() {
+    vm.rows = vm.getSelectedRows();
+  }
+
   // in order to use controllerAs syntax, we need to import the entire grid API
   // into the controller scope to bind the getSelectedRows method.
-  function bindGridApi(api) {
+  function onRegisterApi(api) {
```
Comment (DedrickEnc) [inline client/src/partials/cash/modals/invoices.modal.js:50]: Why you are not using our grid service for selection (`GridSelection`)? that should help us having a good grid selection because it will be more flexible and capable to deal with many type of grid.

File: server/controllers/finance/cash.create.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  // make sure uuids are defined and converted
+  items.forEach(item => {
+    item.cash_uuid = cashUuid;
+    item.uuid = item.uuid || uuid.v4();
+
+    item = db.convert(item, ['uuid', 'invoice_uuid']);
+  });
+
+  // make sure the items are in an ordered array
```
Comment (DedrickEnc) [inline server/controllers/finance/cash.create.js:36]: Is it helpful to order this array?

File: client/src/partials/cash/cash.html (hunk 2 above)
Comment (jniles) [reply client/src/partials/cash/cash.html:113]: Yup!  Good idea.

File: client/src/partials/cash/cash.html (hunk 1 above)
Comment (jniles) [reply client/src/partials/cash/cash.html:55]: Actually, the only dependency here is for an `<a href>`  to link to the correct `/cashboxes/:id/update` URL.  Otherwise, the component doesn't depend on it and it is optional.

File: client/src/partials/cash/cash.js (hunk 3 above)
Comment (jniles) [reply client/src/partials/cash/cash.js:40]: It's still a part - I've just moved it to a separate `ui-router` state.  So, instead of `ng-click="ctrl.openTransferModal()"`, we have `<a ui-sref="cash.transfer(ctrl.cashbox.id)">`.  This change was introduced in a previous PR that dealt with the cash client.

For information, see [cash.routes](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/partials/cash/cash.routes.js#L25).

File: client/src/partials/cash/modals/invoices.modal.js (hunk 4 above)
Comment (jniles) [reply client/src/partials/cash/modals/invoices.modal.js:50]: This modal was written about four months before the `GridSelection` service was written.  I tried to keep my clientside changes minimal in this PR to focus on the server posting logic.  Most of the changes on the client were attempted bug fixes or removing old code.  In a separate PR, I'll come through and finish the client refactor.

File: server/controllers/finance/cash.js [hunk 6]
```diff
@@ -25,14 +25,16 @@ const NotFound = require('../../lib/errors/NotFound');
 const BadRequest = require('../../lib/errors/BadRequest');
 const journal = require('./journal/cash');
```
Comment (DedrickEnc) [inline server/controllers/finance/cash.js:26]: It seems like journal is no longer used, so it should be deleted.

File: server/controllers/finance/cash.create.js (hunk 5 above)
Comment (jniles) [reply server/controllers/finance/cash.create.js:36]: It is completely necessary.  The values are expected to be inserted in a specific order ('uuid', 'cash_uuid', 'amount', 'invoice_uuid') and JSON is an unordered format.  [See this answer](http://stackoverflow.com/a/5525820).

File: server/models/schema.sql [hunk 7]
```diff
@@ -704,13 +704,15 @@ CREATE TABLE `event` (
...
-  `enterprise_id`   smallint(5) UNSIGNED NOT NULL,
+  `enterprise_id`   SMALLINT(5) UNSIGNED NOT NULL,
   `currency_id`   TINYINT(3) UNSIGNED NOT NULL,
   `rate`    DECIMAL(19,4) UNSIGNED NOT NULL,
   `date`    DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
   PRIMARY KEY (`id`),
   KEY `enterprise_id` (`enterprise_id`),
   KEY `currency_id` (`currency_id`),
+  INDEX `rate` (`rate`),
+  INDEX `date` (`date`),
```
Comment (DedrickEnc) [inline server/models/schema.sql:715]: Can we prevent inserting a duplicate date and rate in the exchange rate table? May be in a separate PR
